### PR TITLE
2.6.1 -> 2.7

### PR DESCRIPTION
**Pull request by endofline**
# Deprecation Notices

* The following **[Message Sending](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending)** functions are **soft-deprecated** and will raise `[DEPRECATED]` DEBUG notices in the log if they are used (4fb92c6):
  * ~~`bot.send_message()`~~ - ~~replace with **[bot.coro_send_message()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation)**~~ [*un-deprecated*](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation) in 652901e
  * `bot.send_message_parsed()` - replace with **[bot.coro_send_message()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation)**
  * `bot.send_message_segments()` - replace with **[bot.coro_send_message()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation)**
  * `bot.send_html_to_conversation()` - replace with **[bot.coro_send_message()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation)**
* The following **[Message Sending](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending)** functions are **hard-deprecated** and will raise `[DEPRECATED]` WARNINGS in the logger and on console if they are used:
  * `bot.external_send_message()` - replace with **[bot.coro_send_message()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation)**
  * `bot.external_send_message_parsed()` - replace with **[bot.coro_send_message()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation)**
  * `bot.send_html_to_user()` - replace with **[bot.coro_send_to_user()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-user-via-1-to-1)**
  * `send_html_to_user_or_conversation` - replace with combination of **[bot.coro_send_message()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation)** and **[bot.coro_send_to_user()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-user-via-1-to-1)**
* The [**Pre-2.4 Plugin Pattern**](https://github.com/hangoutsbot/hangoutsbot/wiki/Plugin-Authoring-(2.3.4-and-below))  is **hard-deprecated** and continued usage will raise `[DEPRECATED]` WARNINGS in the logger and on console if they are used. (re: #301, 5712455)
  * Modernised plugin patterns: **Forwarding** (1093a72), **Image Links** (6adfa98), **Mentions** (dfaa9f2), **SimplyTranslate** (8ed1bab), **Slack** (52b52cf), **Subscribe** (1e8922e)
  * **PLUGIN DEVELOPERS: [Please migrate](https://github.com/hangoutsbot/hangoutsbot/wiki/Migrating-Pre-2.4-Plugins)**

# Features

* **[Access Control with Tags](https://github.com/hangoutsbot/hangoutsbot/wiki/Access-Control-with-Tags)** (re: #293)
* **System Plugins** which provides a [richer core command set](https://github.com/hangoutsbot/hangoutsbot/wiki/Core-Commands) for bots (re: #293)
  * Note: If you receive warnings for missing plugin [**unittest_convmem**](https://github.com/hangoutsbot/hangoutsbot/wiki/Permanent-Memory#Permamem_Diagnostics__Repair) or [**unittest_tags**](https://github.com/hangoutsbot/hangoutsbot/wiki/Users-and-Conversations-Tagging#basic-tag-manipulation), please remove the entries from your manually-configured [`config["plugins"]`](https://github.com/hangoutsbot/hangoutsbot/wiki/Configuration#plugins) key, as both plugins have been integrated as system plugins
  * **New Plugin Control Commands** allows load/unload/reload of all plugins - except threaded ones due to platform constraints (b20d29b)
* [**(Python) Logger to Chat Message**](https://github.com/hangoutsbot/hangoutsbot/wiki/Logger-to-Chat-Message) provides simple redirection of log messages to any conversations tagged `receive-logs` (re: #196, 4e3c7cf)
* **`aiohttp.web` sinks** - faster than threaded sinks by order of a magnitude (6492d7f)
  * adds `AsyncRequestHandler` as counterpart to `BaseBotRequestHandler`
  * framework function `sinks.aiohttp_start()` to start aiohttp.web sinks
  * smart sinks loader starts `config[json_rpc]`-configured sink based on parent class `AsyncRequestHandler` or `BaseBotRequestHandler`
  * **webbridge** framework feature updated to use `AsyncRequestHandler` - reduces response time for webbridge-based **Hubot** plugin (376dc61) 
* **Long-running `asyncio.task`**, replaces threads, start a task by registering a long-running coroutine function with `plugins.register_asyncio_task(coroutine_function, [<extra arguments>])` - `bot` is always passed as the first parameter to the registered function

# Plugins and Sinks

* **`AsyncMessagePoster`** is a **NEW SINK** that is a drop-in replacement for `SimpleMessagePoster` - uses `AsyncRequestHandler` for faster response time to incoming requests
* **Restricted Add** - Bugfix: Prevent bot from making itself a botkeeper even if the bot id has been added to `config["admins"]` (re: #303, e87dd30)
* **Namelock** - Enhancement: Update locked topic if authorised user (bot, admin) renames the conversation (re: #304, 22ec152)
* **Slack** - Enhancement: More informative errors on configuration/initialisation errors (re: #306, re: #238, 52b52cf); Re-factored to use `AsyncRequestHandler` (6492d7f)
* **Google Script Sink** - Enhancement: Upgraded to use [BaseBotRequestHandler](https://github.com/hangoutsbot/hangoutsbot/wiki/BaseBotRequestHandler-Class) (re: #297,  51080dc); improved test script (4cbd7fd); new [wiki page] (https://github.com/hangoutsbot/hangoutsbot/wiki/Google-Script-Sink); Refactored to use `AsyncRequestHandler` (4d623bd)
* **Cleverbot** - Bugfix: Fix broken API with code adapted from latest [folz/cleverbot.py](https://github.com/folz/cleverbot.py) (e5e5216); Enhancement: Prevent bot hangs when plugin is busy (re: #187, d7b96ea)
* **BotAlive** - Bugfix: Prevent critical errors in the plugin from quitting the bot (2d5ca57) - also applied as an emergency fix to master via [2.6.1.patch-1 release](https://github.com/hangoutsbot/hangoutsbot/releases/tag/2.6.1.patch-1); Enhancement: Refactored as a long-running `asyncio.task` (d7273fa); Added configurable timing (e6dd08d); [Wiki updated](https://github.com/hangoutsbot/hangoutsbot/wiki/Botalive-Plugin)
* **Syncrooms** - Enhancement: Remove redundant "Incoming image..." (f1f93b1)
* **Forwarding** - Enhancement: Remove redundant "Incoming image..." (396c2d5)
* **TLDR** - Bugfix: Prevent `memory.json` corruption on preloaded entries (b2f230b) - also applied as an emergency fix to master via [2.6.1.patch-2 release](https://github.com/hangoutsbot/hangoutsbot/releases/tag/2.6.1.patch-2); Bugfix: Regression with TLDR sorting introduced with corruption fix (a099e4f via #309, thanks @j16sdiz!)
* **Default** - Bugfix/Enhancement: Fix broken `/bot broadcast remove` sub-command and potential markup mangling in `/bot broadcast info` output (re: #313, 21b3603)
* **API** - Enhancement: Return command output (requires a command to return dictionary containing key `api.response`); Re-factored to use `AsyncRequestHandler` (6492d7f)
  * `ping` (9109bca), `convfilter`, `convusers` (32542ee) modified to return command output
* **Github Push Listener** - Enhancement: Refactored to use `AsyncRequestHandler` (re: #326, 0fce5e9)
* **Gitlab Push Listener** - Enhancement: Refactored to use `AsyncRequestHandler` (re: #326, 0fce5e9)

# Other Changes & Upgrades

* Framework automatically monkey-patches OTR support into hangups for `adduser` and `removeuser` - bot obeys OTR status when adding users and leaving a conversation (re: #302, e2f3263)
* [Force OTR flag](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#force-off-the-record-otr-flag) when sending specific messages by using `context` parameter (re: #299, d7a7272)
* Sinks: More informative errors when `.pem` file is missing (re: #305 + #306, e97e9b4, 5735ee7); address is in use (re: #306, 5735ee7); and when `.pem` file is corrupt or invalid (re: #306, df696d6)
* [`config["autocreate-1to1"]`](https://github.com/hangoutsbot/hangoutsbot/wiki/Configuration#autocreate1to1) is now defaulted ON (re: #202, 76c9848)
* **[bot.coro_send_message()](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation)** implemented as the replacement for `bot.send_message_parsed()`, `bot.send_html_to_conversation()`, `bot.send_message_segments` and `bot.send_message` (90df3fd, 9811441, 375d8f0)
  * Older functions will continue to work but will be deprecated and obsoleted in upcoming versions. Plugin developers are encouraged to update their older patterns since this is a drop-in replacement.
  * Plugins updated: **Image Links** (873a337), **Core: Basic** (01de566), **Namelock** (30cd891), **Autoreply** (3ad1a46), **Cleverbot** (3ad1a46), **Reddit Image Linker** (3ad1a46, b8b75b9), **Lookup** (3ad1a46), **MonitorAdds** (3ad1a46), **unittest_geticon** (3ad1a46), **WolframAlpha** (3ad1a46), **TLDR** (7728df3), **Meme Generator Images** (7728df3, b8b75b9), **Forwarding** (7728df3), **Conversation Tools** (7fb9256), **Conversation Invitations** (7fb9256), **Image Screenshot** (7fb9256), **Syncrooms** (f1f93b1), **Default** (727c3ec), **Forwarding** (396c2d5), **Botalias** (c6c4f63), **Chance** (c6c4f63), **DND** (c6c4f63), **Hangout Calls Humor** (c6c4f63), **Syncroom Autotranslate** (c6c4f63), **Syncroom Config** (c6c4f63), **Memory Example** (0a68b65), **Lottery** (0a68b65), **SimpleWikipedia** (0a68b65), **SimplyTranslate** (0a68b65), **Starter** (0a68b65), **Urbandict** (0a68b65), **Mentions** (b720d1a), **Subscribe** (b720d1a), **Restricted Add** (b720d1a), **Core: CONVID** (0c39d68), **Core: Permamem** (0c39d68), **Core: Tagging** (0c39d68), **unittest_reprocessor** (892d95c), **Slack** (cc73294), **API** (cc73294)
  * Sinks updated: **gitlab** (cc73294), **github** (cc73294), **Google Script Sink** (cc73294), **simpledemo** (fc8582e)
  * Framework Components Updated: **CommandDispatcher** (0c39d68), *event handling* (892d95c), **BaseBotRequestHandler** (fc8582e)
* **Reprocessor** supports asyncio.coroutine or conventional function as mapped function - **unittest_reprocessor** updated to demonstrate both mechanisms (892d95c)
* Direct API call to `bot._client.removeuser()` instead of OOP fluff in `convleave` (107601a) 
* Command Dispatcher will now try and match case-insensitive command if it cannot find a case-sensitive matching command (re: #314, 46f4c26)
* Additional plugins updated to reflect python logging convention (re: #319, 7a58dc6, thanks @pleasantone!)
* All unittests and example plugins are deactivated on default configuration (re: #320, 947a052, per @pleasantone's suggestion)
* Override the default python logging settings by bootstrapping config (re: #321, #324, 55e20b9, thanks @pleasantone)

# Documentation

* *PEM Generation* instructions separated moved to [new wiki page](https://github.com/hangoutsbot/hangoutsbot/wiki/PEM-File-Generation)
* *Google Scripts Sink setup and configuration* moved to [new wiki page](https://github.com/hangoutsbot/hangoutsbot/wiki/Google-Script-Sink)
* Page conversion from GFM to Mediawiki format (for automated Table-of-Contents generation):
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Access-Control-with-Tags
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Configuration
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Customising-Access-Control
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Permanent-Memory
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Role-Based-Access-System-Discussion
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Users-and-Conversations-Tagging
* Page renames for compatibility with Window:
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Reference:-BaseBotRequestHandler-Class to **https://github.com/hangoutsbot/hangoutsbot/wiki/BaseBotRequestHandler-Class**
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Reference:-hangups-state_update-dumps to **https://github.com/hangoutsbot/hangoutsbot/wiki/hangups-state_update-dumps**
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Reference:-Manipulating-config.json to **https://github.com/hangoutsbot/hangoutsbot/wiki/Accessing-and-Writing-config.json**
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Reference:-Permanent-Conversation-Memory, https://github.com/hangoutsbot/hangoutsbot/wiki/Permanent-Conversation-Memory, https://github.com/hangoutsbot/hangoutsbot/wiki/Reference:-Permanent-Memory-(permamem), https://github.com/hangoutsbot/hangoutsbot/wiki/Permanent-Memory-(permamem)
 to **https://github.com/hangoutsbot/hangoutsbot/wiki/Permanent-Memory**
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Reference:-Manipulating-memory.json to **https://github.com/hangoutsbot/hangoutsbot/wiki/Accessing-and-Writing-memory.json**
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Reference:-Sending-Messages to **https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending**
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Reference:-Tags-List to **https://github.com/hangoutsbot/hangoutsbot/wiki/Tags-List**
  * https://github.com/hangoutsbot/hangoutsbot/wiki/Reference:-User-and-Conversation-Tagging to **https://github.com/hangoutsbot/hangoutsbot/wiki/Users-and-Conversations-Tagging**

### CODE REVIEW

**Comment by PaulTimmins** [comment]
Is it really necessary to deprecate bot.send_message in favor of the less memorable drop in replacement bot.coro_send_message? Is it harmful to leave a permanent map?

**Comment by endofline** [comment]
@PaulTimmins - I was actually trying to deal with the proliferation of message sending functions in the core framework. There isn't a problem keeping one or two of these within the framework, and they could act as aliases to the `coro_` function where possible, somewhat in the [same way that send_message_segments()](https://github.com/hangoutsbot/hangoutsbot/blob/staging/hangupsbot/hangupsbot.py#L308) is right now in staging. However, the original functionality of a function like `send_message()` might change since previously it sent non-formatted text (because of historical reasons associated with the original bot and the hangups client library), whereas the present parses the incoming text for html and markdown.

**Comment by endofline** [comment]
"Un-deprecated" `bot.send_message()` keeping its original behaviour of sending plain-text messages by default. Added optional parser toggle override via `context` variable  and image support to "modernise" the whole function. Updated [standard messaging function documentation](https://github.com/hangoutsbot/hangoutsbot/wiki/Message-Sending#messaging-arbitrary-conversation). Usage of coroutine-based `coro_send_message()` is still recommended though since this is a design consideration/constraint from the hangups client library.
